benchmark/paddle/CARLA_SAC/train.py

```diff
@@ -0,0 +1,153 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import gym
+import numpy as np
+import parl
+import argparse
+import carla
+import gym_carla
+from env_utils import ParallelEnv, LocalEnv
+from parl.utils import logger, tensorboard, ReplayMemory
+from parl.env.continuous_wrappers import ActionMappingWrapper
+from carla_model import CarlaModel
+from carla_agent import CarlaAgent
+from sac import SAC
+# from parl.algorithms import SAC # parl >= 1.4.2
+from env_config import EnvConfig
+
+WARMUP_STEPS = 2e3
+EVAL_EPISODES = 3
+MEMORY_SIZE = int(1e4)
+BATCH_SIZE = 256
+GAMMA = 0.99
+TAU = 0.005
+ALPHA = 0.2  # determines the relative importance of entropy term against the reward
+ACTOR_LR = 3e-4
+CRITIC_LR = 3e-4
+
+
+# Runs policy for 3 episodes by default and returns average reward
+def run_evaluate_episodes(agent, env, eval_episodes):
+    avg_reward = 0.
+    for k in range(eval_episodes):
+        obs = env.reset()
+        done = False
+        steps = 0
+        while not done and steps < env._max_episode_steps:
+            steps += 1
+            action = agent.predict(obs)
+            obs, reward, done, _ = env.step(action)
+            avg_reward += reward
+    avg_reward /= eval_episodes
+    return avg_reward
+
+
+def main():
+    logger.info("-----------------Carla_SAC-------------------")
+    logger.set_dir('./{}_train'.format(args.env))
+
+    # Parallel environments for training
+    train_envs_params = EnvConfig['train_envs_params']
+    env_list = ParallelEnv(args.env, args.xparl_addr, train_envs_params)
+
+    # env for eval
+    eval_env_params = EnvConfig['eval_env_params']
+    eval_env = LocalEnv(args.env, eval_env_params)
+
+    obs_dim = eval_env.obs_dim
+    action_dim = eval_env.action_dim
+
+    # Initialize model, algorithm, agent, replay_memory
+    model = CarlaModel(obs_dim, action_dim)
+    algorithm = SAC(
+        model,
+        gamma=GAMMA,
+        tau=TAU,
+        alpha=ALPHA,
+        actor_lr=ACTOR_LR,
+        critic_lr=CRITIC_LR)
+    agent = CarlaAgent(algorithm)
+    rpm = ReplayMemory(
+        max_size=MEMORY_SIZE, obs_dim=obs_dim, act_dim=action_dim)
+
+    total_steps = 0
+    last_save_steps = 0
+    test_flag = 0
+
+    obs_list = env_list.reset()
+
+    while total_steps < args.train_total_steps:
+        # Train episode
+        if rpm.size() < WARMUP_STEPS:
+            action_list = [
+                np.random.uniform(-1, 1, size=action_dim)
+                for _ in range(len(train_envs_params))
+            ]
+        else:
+            action_list = [agent.sample(obs) for obs in obs_list]
+        next_obs_list, reward_list, done_list, info_list = env_list.step(
+            action_list)
+
+        # Store data in replay memory
+        for i in range(len(train_envs_params)):
```

Comment:
Do not repeatedly call `len(train_envs_params)` to access the number of environments.